### ISSUE TITLE
Build issue when using libc++ as standard library

### ISSUE THREAD

**Issue by Wentzell** [krivenko/libcommute#5]
Dear @krivenko,

When I build libcommute cad615dbe56ac34e28dfbcf870565e5ac469f9c6 with clang-15 using libc++ (`-stdlib=libc++`) I get the following compilation issue


```
In file included from /mnt/home/wentzell/Dropbox/Coding/libcommute/examples/gamma.expression.cpp:18:                                                                                                                                   In file included from /mnt/home/wentzell/Dropbox/Coding/libcommute/examples/gamma.hpp:118:
In file included from /mnt/home/wentzell/Dropbox/Coding/libcommute/include/libcommute/expression/expression.hpp:20:
/mnt/home/wentzell/Dropbox/Coding/libcommute/include/libcommute/expression/monomial.hpp:153:12: error: no matching conversion for functional-style cast from 'std::vector<std::unique_ptr<libcommute::generator<int>>>::const_reverse_i
terator' (aka 'reverse_iterator<__wrap_iter<const std::unique_ptr<libcommute::generator<int>, std::default_delete<libcommute::generator<int>>> *>>') to 'libcommute::monomial<int>::const_reverse_iterator' (aka 'reverse_iterator<libcommute::monomial<int>::const_iterator>')
    return const_reverse_iterator(generators_.rbegin());                                                                                                                                                                                          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~                                                                                                                                                                                 /mnt/home/wentzell/Dropbox/Coding/libcommute/include/libcommute/expression/expression.hpp:197:33: note: in instantiation of member function 'libcommute::monomial<int>::rbegin' requested here
        for(auto m_it = m.first.rbegin(); m_it != m.first.rend(); ++m_it) {
                                ^
/mnt/home/wentzell/Dropbox/Coding/libcommute/examples/gamma.expression.cpp:53:56: note: in instantiation of member function 'libcommute::conj' requested here                                                                            std::cout << "gamma5 - conj(gamma5) = " << (gamma5 - conj(gamma5))                                                                                                                                                                                                                          ^
/mnt/home/wentzell/opt/llvm_15.0.3/bin/../include/c++/v1/__iterator/reverse_iterator.h:43:28: note: candidate constructor (the implicit copy constructor) not viable: no known conversion from 'reverse_iterator<std::vector<std::uniqu
e_ptr<libcommute::generator<int>>>::const_iterator>' to 'const reverse_iterator<libcommute::monomial<int>::const_iterator>' for 1st argument
class _LIBCPP_TEMPLATE_VIS reverse_iterator
                           ^
/mnt/home/wentzell/opt/llvm_15.0.3/bin/../include/c++/v1/__iterator/reverse_iterator.h:43:28: note: candidate constructor (the implicit move constructor) not viable: no known conversion from 'reverse_iterator<std::vector<std::uniqu
e_ptr<libcommute::generator<int>>>::const_iterator>' to 'reverse_iterator<libcommute::monomial<int>::const_iterator>' for 1st argument
/mnt/home/wentzell/opt/llvm_15.0.3/bin/../include/c++/v1/__iterator/reverse_iterator.h:88:14: note: candidate constructor not viable: no known conversion from 'std::vector<std::unique_ptr<libcommute::generator<int>>>::const_reverse
_iterator' (aka 'reverse_iterator<__wrap_iter<const std::unique_ptr<libcommute::generator<int>, std::default_delete<libcommute::generator<int>>> *>>') to 'libcommute::monomial<int>::const_iterator' for 1st argument
    explicit reverse_iterator(_Iter __x) : __t(__x), current(__x) {}
             ^
/mnt/home/wentzell/opt/llvm_15.0.3/bin/../include/c++/v1/__iterator/reverse_iterator.h:94:5: note: candidate template ignored: requirement 'is_convertible<const std::__wrap_iter<const std::unique_ptr<libcommute::generator<int>, std
::default_delete<libcommute::generator<int>>> *> &, libcommute::monomial<int>::const_iterator>::value' was not satisfied [with _Up = std::__wrap_iter<const std::unique_ptr<libcommute::generator<int>> *>]
    reverse_iterator(const reverse_iterator<_Up>& __u)
    ^
/mnt/home/wentzell/opt/llvm_15.0.3/bin/../include/c++/v1/__iterator/reverse_iterator.h:85:5: note: candidate constructor not viable: requires 0 arguments, but 1 was provided
    reverse_iterator() : __t(), current() {}
    ^
In file included from /mnt/home/wentzell/Dropbox/Coding/libcommute/examples/gamma.expression.cpp:18:
In file included from /mnt/home/wentzell/Dropbox/Coding/libcommute/examples/gamma.hpp:118:
In file included from /mnt/home/wentzell/Dropbox/Coding/libcommute/include/libcommute/expression/expression.hpp:20:
```

**Comment by krivenko** [comment]
Hi Nils,

Please, check if the latest commit on branch `issue5` fixes your compilation problem (it does for me).

**Comment by Wentzell** [comment]
Hi Igor,

The issue is resolved when I use the `issue5` branch.

**Comment by krivenko** [comment]
Great, the fix ([7fe4a01](https://github.com/krivenko/libcommute/commit/7fe4a01920d032fcab33a5f021630c1eefa436b4)) is on master now.

**Comment by Wentzell** [comment]
Thank you @krivenko
